File: leetcode/src/00206_reverse_linked_list/solution.go
package leetcode

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	cur := head
	for cur != nil {
		next := cur.Next
		cur.Next = prev
		prev = cur
		cur = next
	}
	return prev
}

func solve(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	newHead := solve(head.Next)
	head.Next.Next = head
	head.Next = nil
	return newHead
}

File: leetcode/src/00206_reverse_linked_list/solution.rs
struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut prev, mut cur) = (None, head);
        while let Some(mut node) = cur {
            cur = node.next;
            node.next = prev;
            prev = Some(node)
        }
        prev
    }

    fn solve(head: Option<Box<ListNode>>, prev: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        match head {
            Some(mut node) => {
                let cur = node.next.take();
                node.next = prev;
                Solution::solve(cur, Some(node))
            }
            None => prev,
        }
    }
}
